WIP: Waiting for DBFTables.write

joshday opened the pull request:
Adds .dbf files when `Shapefile.write`-ing stuff.

rafaqz commented on src/writer.jl:
```diff
@@ -22,17 +22,27 @@ function write(path::AbstractString, obj; force=false)
+        tbl = Dict(field => Tables.getcolumn(obj, field) for field in sch.names)
+        # find first column that only contains geometries (and possibly missing)
+        geomcol = findfirst(tbl) do data
+            all(x -> ismissing(x) || GI.geomtrait(x) isa GI.AbstractGeometryTrait, data) && any(!ismissing, data)
```
```suggestion
            all(x -> ismissing(x) || GI.isgeometry(x, data) && any(!ismissing, data)
```

This is a little simpler.

rafaqz commented on src/writer.jl:
```diff
@@ -22,17 +22,27 @@ function write(path::AbstractString, obj; force=false)
         geoms = (GI.getgeom(obj, i) for i in 1:GI.ngeom(obj))
+        tbl = fill((;feature=missing), length(geoms))  # fill .dbf with missing so that it exists
+    elseif GI.geomtrait(obj) isa GI.AbstractFeatureCollectionTrait
+        tbl = GI.properties.(obj)
```
I don't think you can assume broadcast will work here. But this should:
```suggestion
    tbl = [properties(f) for f in getfeature(obj)]
```

rafaqz commented on src/writer.jl:
```diff
@@ -22,17 +22,27 @@ function write(path::AbstractString, obj; force=false)
+        tbl = fill((;feature=missing), length(geoms))  # fill .dbf with missing so that it exists
+    elseif GI.geomtrait(obj) isa GI.AbstractFeatureCollectionTrait
+        tbl = GI.properties.(obj)
+        geoms = GI.geometry.(obj)
```
```suggestion
    geoms = [geometry(f) for f in getfeature(obj)]
```

rafaqz commented on src/writer.jl:
```diff
@@ -22,17 +22,27 @@ function write(path::AbstractString, obj; force=false)
+        # tbl = features
     else
         geoms = obj
+        tbl = fill((;feature=missing), length(geoms))
```
Is the idea that there would be an empty `feature` column if there is no data for the table?

Can we not just have a totally empty table?

joshday replied:
> The dBASE table contains feature attributes with one record per feature. The one-to-one relationship between geometry and attributes is based on record number. Attribute records in the dBASE file must be in the same order as records in the main file.

- ESRI Shapefile Technical Description: https://www.esri.com/content/dam/esrisites/sitecore-archive/Files/Pdfs/library/whitepapers/pdfs/shapefile.pdf

rafaqz replied:
Ok makes sense

evetion replied:
But we could just *not* write the file in case of an empty table right?

rafaqz replied:
Yeah that's another option. `Table` doesn't wirk without a .dbf present. It should at least round trip properly so we could fix the `isfile(paths.dbf)` check instead to give `@info` rather than an error, and let the dbf field be `nothing`.

evetion replied:
Reading the spec it seems we must write a table, so this should work.

evetion commented on src/writer.jl:
```diff
@@ -22,17 +22,23 @@ function write(path::AbstractString, obj; force=false)
+        sch = Tables.schema(obj)
+        tbl = Dict(field => Tables.getcolumn(obj, field) for field in sch.names)
+        # find first column that only contains geometries (and possibly missing)
+        geomcol = findfirst(tbl) do data
```
The geomcol = first(GI.geometrycolumns(obj)) is the idiomatic way to check this, where geometrycolumns defaults to a `(:geometry,)`, but one should indeed check isgeometry afterwards.

Besides, there *could* be multiple geometry columns, we only take the first, but we should pop the remainder as well, because I don't expect DBF to like them much.

rafaqz replied:
DBF will hate them

joshday replied:
> The geomcol = first(GI.geometrycolumns(obj)) is the idiomatic way to check this

But say for DataFrames, this will just return `:geometry` every time.  Isn't it better to check the column types?

rafaqz replied:
That's a good point.

Can we first check the columnnanes, and check the types as a fallback?

joshday replied:
That leaves an edge case that will error if you have a `:geometry` column that doesn't contain geometries, but I'll defer to y'all.

joshday commented:
Okay, I just rewrote a significant portion.  I've added a `Shapefile.Writer` struct that provides low-level usage if you want to provide your own iterator of `geoms`, a `table`, and a `crs`.  Every `Shapefile.write` method should pass through it.


`Shapefile.write(path, obj)` works for any `obj` that satisfies:

1. `isgeometry(obj)`
2. `all(isgeometry, obj)`
3. `isfeature(obj)`
4. `all(isfeature, obj)`
5. `Tables.istable(obj)` and at least one columns contains geoms satisfying `isgeometry(element)`
6. `isfeaturecollection(obj)`
7.  `geomtrait(obj) <: AbstractGeometryCollectionTrait`

Currently `missing` shapes are dropped, but that can be changed once null shapes (https://github.com/JuliaGeo/Shapefile.jl/issues/81) are added.

My test cases could use review to make sure I'm using GeoInterface correctly.

joshday commented:
Never mind on `missing`.  That will be supported in a minute...
